Feature: Theory of Change page layout enhancements

MalloryLeeWong opened the pull request:
### Background

This improves the Theory of Change page layout and styling for https://github.com/Blue-Marine-Foundation/Maerl/issues/137 

### Changes
- Removed information about Outcome Indicators
- Swapped layout from horizontal to vertical
- Updated the main fetchLogframe server action (in the @/components/logframe/server-actions.ts file) to fetch impact, outcome, and outputs as siblings (currently they are parent/child/grandchild)
- Listed outputs as cards underneath the Outcome
- Removed "Edit" buttons
- Enhanced colors and spacing of the cards to more easily distinguish them visually
- Reduced size of the Impact and Outcome cards

### Screenshots

<img width="1620" alt="TOC-pr-feedback-yavin4-left-card-titles" src="https://github.com/user-attachments/assets/41e73250-7105-447c-9b2a-67150121c2e7" />

MalloryLeeWong commented:
@samjamead FYI put my work so far on the TOC layout improvements into this draft PR. The major changes are done, but tomorrow will polish off a few details then will let you know when it's ready for final review.

MalloryLeeWong commented:
@samjamead This PR is ready for your review when you have time. Appreciate any feedback! Aimed to refactor without changing the existing component and data passing/prop patterns too much. Beyond what we discussed during our last meeting, I also attempted to enhance the colors since Sophie suggested something like this in the wireframe videos.

I created a branch off of main `[feat/theory-of-change-and-logframe-view-enhancements](https://github.com/Blue-Marine-Foundation/Maerl/tree/feat/theory-of-change-and-logframe-view-enhancements)` that we can use for all of the TOC and Logframe improvements. I am pointing this PR's feature branch to there.

samjamead reviewed (commented):
Hey @MalloryLeeWong thanks so much for your work so far, I really like the direction that you're taking things! 

I have pulled the code from the feature branch and just took a quick look. A few thoughts spring to mind. 

(One thing I think it's worth flagging is that some of my suggestions here are relevant to code that I have written previously. So please don't take these as negative feedback for your work, rather, me just taking the opportunity to review my own code and suggest simplifications and improvements that we can take the opportunity to implement now while you're working on things!)

- I always like to limit the width of text using the TW class `max-w-prose` to maximise readability.
- Once the text is limited in width the cards have a lot of empty space to the right so I wonder if it makes sense to shrink them down to `max-w-fit`. Having said that, I know Sophie's sketches have the cards as full width, so perhaps we can play around with these two options (this is actually one of the reasons that I went for a column layout previously, so we could always revisit that if Sophie decides she doesn't like the full width layout when she sees it in practice).
- [Self review of bad practice] I can see that the ToC page is a shell that contains only one component, the ToC component. I think it makes sense to remove this component and move its logic and layout to the page level. 
- I love how you've added some colour to the page! I wonder if the highlight colours would work better behind the header, with the body of the card a more neutral background colour? 
- [Another point I'm picking up on myself] I notice that the Outputs are sorted alphabetically, so for a project with more than 9 outputs (e.g. Jersey) they are listed as 1, 10, 2, 3, etc. We should update the sorting method to use the Output number. 
- Similarly, I wonder if it makes more sense to label the Output cards as, e.g., 'Output 1', rather than having the generic title and then displaying the code within the card. 
- Now that we're removing the nesting relationship between Outcomes, Outcome Indicators and Outputs, I think it makes sense to also remove the nesting of the Output components from within the `outcome-card` component. 

Here's a quick artists impression to illustrate my suggestions. Let me know what you think and please don't consider these suggestions obligatory. Rather, they are some pointers to bear in mind as options when presenting to Sophie. 

![Screenshot 2025-01-30 at 21 07 56](https://github.com/user-attachments/assets/76285233-faf3-471f-9113-ad1662aef250)

MalloryLeeWong commented:
> Hey @MalloryLeeWong thanks so much for your work so far, I really like the direction that you're taking things!
> 
> I have pulled the code from the feature branch and just took a quick look. A few thoughts spring to mind.
> 
> (One thing I think it's worth flagging is that some of my suggestions here are relevant to code that I have written previously. So please don't take these as negative feedback for your work, rather, me just taking the opportunity to review my own code and suggest simplifications and improvements that we can take the opportunity to implement now while you're working on things!)
> 
> * I always like to limit the width of text using the TW class `max-w-prose` to maximise readability.
> * Once the text is limited in width the cards have a lot of empty space to the right so I wonder if it makes sense to shrink them down to `max-w-fit`. Having said that, I know Sophie's sketches have the cards as full width, so perhaps we can play around with these two options (this is actually one of the reasons that I went for a column layout previously, so we could always revisit that if Sophie decides she doesn't like the full width layout when she sees it in practice).
> * [Self review of bad practice] I can see that the ToC page is a shell that contains only one component, the ToC component. I think it makes sense to remove this component and move its logic and layout to the page level.
> * I love how you've added some colour to the page! I wonder if the highlight colours would work better behind the header, with the body of the card a more neutral background colour?
> * [Another point I'm picking up on myself] I notice that the Outputs are sorted alphabetically, so for a project with more than 9 outputs (e.g. Jersey) they are listed as 1, 10, 2, 3, etc. We should update the sorting method to use the Output number.
> * Similarly, I wonder if it makes more sense to label the Output cards as, e.g., 'Output 1', rather than having the generic title and then displaying the code within the card.
> * Now that we're removing the nesting relationship between Outcomes, Outcome Indicators and Outputs, I think it makes sense to also remove the nesting of the Output components from within the `outcome-card` component.
> 
> Here's a quick artists impression to illustrate my suggestions. Let me know what you think and please don't consider these suggestions obligatory. Rather, they are some pointers to bear in mind as options when presenting to Sophie.
> 
> ![Screenshot 2025-01-30 at 21 07 56](https://private-user-images.githubusercontent.com/3154415/408299526-76285233-faf3-471f-9113-ad1662aef250.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgyNzM0NjUsIm5iZiI6MTczODI3MzE2NSwicGF0aCI6Ii8zMTU0NDE1LzQwODI5OTUyNi03NjI4NTIzMy1mYWYzLTQ3MWYtOTExMy1hZDE2NjJhZWYyNTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMzBUMjEzOTI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWY2OTc1MWE4NGU4YzE4YjU3NzUwZTRjNTcwOTUzOTAzZWM3NzcwNjE2NWU4OGJjYzUwMGIzMzc0YzM4MDliMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.5IloxAkrurcgxaJwM32_UsAk2Y5ucy8bOS9LruqtQtw)

@samjamead Thank you for your feedback! Great ideas and suggestions - will implement them.

For the card widths, building off your suggestion, another solution could be to limit the text and card width like you did in the screenshot, but also center the cards on the page instead of having them start on the left. Perhaps this could make the excess empty space feel more balanced. I will experiment with the spacing get back to you.

Appreciate your thoughts on how much code simplification to do while we're refactoring now. Agree that there are opportunities to further remove component nesting on this page that's no longer needed due to the new layout, so I'll go ahead and clean that up more.

MalloryLeeWong commented:
@samjamead This PR is ready for your review again. I addressed your feedback. Thanks again for your comments! 

Please see new screenshots in the PR description. 

To streamline the spacing, I implemented your suggestion to use `max-w-prose` for the text, but keep the cards full width, then moved the output status to the right side so the text in every card is left aligned which could make it easier to read. I feel like this could be a relatively clean solution. Interested in your thoughts.

samjamead commented on app/(protected)/projects/[slug]/theory-of-change/page.tsx:
```diff
@@ -1,14 +1,56 @@
+  });
+
+  if (isLoading) {
+    return <TheoryOfChangeSkeleton />;
```
Love the addition  ❤️

samjamead commented on components/logframe/extractOutputCodeNumber.ts:
Good call, I think this will be useful in many places

samjamead reviewed (approved):
Thanks very much Mallory, as well as meeting the brief from Sophie this PR is also a really nice cleanup job on the somewhat messy component tree that existed. Thank you! 

From a code and UI perspective this PR is approved, let's see what kind of feedback Sophie comes up with! 

The GitHub UI has given me some ideas for how we could soak up some of the horizontal space. I'll do a quick mockup and if you like it feel free to show it to Sophie as an option for a future iteration.

samjamead commented:
Something like this is what I am imagining (from my idea above). The side nav could link to the relevant output via smooth scroll. This could be handy on long projects like e.g. Italy. 

![Screenshot 2025-02-01 at 21 05 34](https://github.com/user-attachments/assets/3bf60f1a-e019-49de-b4e1-862f52e620a5)

And then as another iteration while I was playing around, I wonder if shifting the card headers to the left side of the card is a nice way to soak up a bit more horizontal space... The effect I ended up with was something like this: 

![Screenshot 2025-02-01 at 21 17 17](https://github.com/user-attachments/assets/d38ed3a3-4e02-4dde-8cfe-369b94063d60)

I'm not sure I love it... I do like that the status badges are a bit more involved. But I'm not sold on the overall look. 

Anyway, just playing around and throwing some ideas out there. Please feel free to present what you think is best to Sophie!

MalloryLeeWong commented:
Thanks @samjamead for your comments and ideas on how we could further utilize the space! 

Pushed a new commit that implements your idea to move the card titles to the left of each card to better utilize the horizontal space. I really like how it also streamlines the reader's experience so they can read from left to right on each card easily to find the info they need. Here's how it looks now:

<img width="1620" alt="TOC-pr-feedback-yavin4-left-card-titles" src="https://github.com/user-attachments/assets/b3ec785d-a2fc-4788-bd32-03f370c83133" />


It's a great point about how for longer projects with many outputs it could be useful to have additional nav / linking support. Do you know how many outputs a project will typically have on average? If it's a read-only page and most projects will likely have fewer than 15 or so outputs, currently it feels to me like the user wouldn't have to scroll down too much, and without the nav the UI remains simpler, potentially minimizing cognitive load when reading the page.

Agree it will be helpful to get Sophie's feedback and share other options with her and continue refining the page as needed in future iterations.

Please let me know if you have any feedback on the new changes before I merge. Thanks!

MalloryLeeWong commented on components/logframe/feature-card-theory-of-change.tsx:
```diff
@@ -0,0 +1,46 @@
+import { cn } from '@/utils/cn';
```
FYI - To implement the left-aligned card title sections on the TOC page, instead of extending `FeatureCard`, I created a new `FeatureCardTheoryOfChange` component  since the card layout for the TOC page has evolved to be very distinct from the existing `FeatureCard`'s layout. 

I also reverted the changes I had previously made to the `FeatureCard` component so it is left unchanged.

samjamead replied:
Good call!

samjamead reviewed (approved):
All looking great Mallory, thanks for your work on this PR! 

When you demo with Sophie and Suneha it might be worth having the various screenshots on hand to show progress and alternative options, just so they can see why we thought it was best to create the horizontal headers layout

MalloryLeeWong commented:
> All looking great Mallory, thanks for your work on this PR!
> 
> When you demo with Sophie and Suneha it might be worth having the various screenshots on hand to show progress and alternative options, just so they can see why we thought it was best to create the horizontal headers layout

Thanks Sam, agreed, great point. I will have screenshots of various layouts on hand for this during the demo.
